### PR TITLE
Implication graph

### PR DESCRIPTION
**Pull request by jtschuler**
NodeKey, Node, and Graph implementations are all complete, with unit tests.

We could probably stand to be a little stricter with our unit tests on cycle detection (see ImplicationGraphTests.cpp), but this is the initial implementation of all necessary data structures for statically checking routines.

### CODE REVIEW

**Comment by jtschuler** [comment]
Note that the way to add nodes is to add edges between them. This is intentional to reduce memory usage. When a node has no more incoming or outgoing nodes, it is deleted.

Next steps:
- Add list of incoming edges to graph in addition to the outgoing edges
- Add support for implication graph specific operations. E.g. automatically adding contrapositive implication edges in the graph for each edge added. We can either add this directly to our current graph or derive from it for better encapsulation.
- **Static Routine Checking:** Implement graph coloring solution to determine if a routine satisfies safety. NOTE: I think we need to spend more time verifying that our algorithm is actually correct (details below)
- **Routine Conflict Checking:** Once that's done, add a data structure that maintains each routine's mapping from node to color, as well as functions to merge graph colorings for efficient conflict detection. This will look a little different from Static Routine Checking since we want ALL nodes the routine touches (no removals!), and a separate color for the head of each dependency chain).

**Static Routine Checking algorithm** (we need to verify this is correct):
For each command in the routine:
    Find the corresponding node.
    Check adjacent (outgoing) nodes to make sure all nodes are colored. If any are not, return error
    Find node's negation.
    Check adjacent (incoming) nodes to make sure all nodes are NOT colored. If any are, return error

**Routine Conflict Checking algorithm** (we need to verify this is correct too):
All routines maintain their own mapping of nodes to colors
Initially all routines map all nodes to NONE

For each routine:
    for each command:
        find corresponding node, color with HEAD
        Color adjacent (outgoing nodes) with CHAIN (i.e. update previous heads)

For all pairs of routines (we can optimize by merging as we go):
    For all nodes:
        if routine 1 has colored the node and routine 2 has colored its negation, and at least one of those colors is CHAIN, return conflict

**Comment by jtschuler** [comment]
Test

File: test/Knox/Compiler/Graph/NodeTests.cpp
Comment (RMenezes782) [inline test/Knox/Compiler/Graph/NodeTests.cpp:1]: NodeAddEdge could test for single directed edge; e.g. make sure that node2 cannot see node1

File: test/Knox/Compiler/Graph/ImplicationGraphTests.cpp
Comment (RMenezes782) [inline test/Knox/Compiler/Graph/ImplicationGraphTests.cpp:1]: Add unit test for case where the node loses its last outgoing edge but still has at least one incoming edge. In this case the original node should still exist.

Comment (RMenezes782) [inline test/Knox/Compiler/Graph/ImplicationGraphTests.cpp:1]: Add unit test for a cross-edge non cycle

**Review by RMenezes782 — APPROVED** [review]
LGTM

**Pull request merged**